### ISSUE TITLE
checkout_externals should proceed when untracked files are present

### ISSUE THREAD

**Issue by billsacks** [ESMCI/manage_externals#51]
### Summary of Issue:

When untracked files are present in one of the externals managed by manage_externals, checkout_externals refuses to proceed since that external is in a "modified" state. I think it's important that it proceed in this case, because it will be very common for users to have untracked files in their directories.

### Expected behavior and actual behavior:

Expected: checkout_externals proceeds despite there being untracked files in one or more external directories.

Actual: checkout_externals bails with a message like this:

```
Processing externals description file : CESM.cfg
Checking status of components: clm, mosart, ww3, cime, cice, pop, cism, rtm, cam, cime_config,
 M  ./cime
    ./cime_config
    ./components/cam
    ./components/cice
    ./components/cism
    ./components/clm
    ./components/mosart
    ./components/pop
    ./components/rtm
    ./components/ww3
----------------------------------------------------------------------
Some external repositories that are not in a clean state. Please
ensure all external repositories are clean before updating.
----------------------------------------------------------------------
```

### Steps to reproduce the problem (should include model description file(s) or link to publi c repository):

1. Clone the cesm repo (`git clone git@github.com:ESCOMP/cesm.git`)

2. Run `./manage_externals/checkout_externals`

3. Add an untracked file in one of the externals directories. In my case, I did a workflow that will be very common: creating a case in `cime/scripts`: `./create_newcase --case mytest --compset A --res f45_g37_rx1`

4. Rerun `./manage_externals/checkout_externals`

### What is the changeset ID of the code, and the machine you are using:

code: cesm repo at a9702c9

Machine: mac (roo2)

### have you modified the code? If so, it must be committed and available for testing:

No

### Screen output or log file showing the error message and context:

See above

**Comment by billsacks** [comment]
@bandre-ucar @mvertens I think it's important to resolve this issue before beta09: I think this will bite a lot of users.

**Comment by billsacks** [comment]
Having now played with manage_externals a lot more, this is the one major issue that I keep running into. It should definitely be fixed before beta09.

**Comment by mvertens** [comment]
@billsacks - I completely agree. I think it should be somewhat
straightforward to just determine if there are untracked files. Do you
agree?

**Comment by billsacks** [comment]
@mvertens probably but I haven't looked at the code.

**Comment by mvertens** [comment]
```
Processing externals description file : CESM.cfg
Checking status of components: clm, mosart, ww3, cime, cice, pop, cism, rtm, cam, cime_config,
See the --help command for meaning of the status output
s   ./cime
    ./cime_config
    ./components/cam
    ./components/cice
    ./components/cism
 M  ./components/clm
    ./components/mosart
    ./components/pop
    ./components/rtm
    ./components/ww3
----------------------------------------------------------------------
The external repositories labeled with M are not in a clean state:

Following are two options for how to proceed:
1) Go into each directory that is not in a clean state, and issue either an 'svn status' or a 'git status' command
  a) if the file has already been committed previously, either revert or commit your changes
  b) if the file is not tracked, then your choices are to move it somewhere else, remove it or commit it

2) Alternatively, you do not have to rely on manage_externals and can
   manually update out of sync externals (labeled with s) as described in the configuratin file, CESM.cfg
```
The above is one suggestion for a more clear error resolution.

**Comment by billsacks** [comment]
Some notes from the discussion between myself, @bandre-ucar and @mvertens (which led to the above comment by @mvertens )

It's usually safe to do a `git checkout` in the presence of untracked files, but not always: it could be that the branch you're checking out contains the given file, in which case git will do the safe thing and refuse to checkout that branch. The fundamental problem here is that we have the requirement that we must either update all externals or update nothing. So we can't say, "Okay, there are untracked files, but I'm going to proceed with the checkout anyway", because this could lead to a situation where some externals were updated successfully but some were not.

I asked if there's something like `--dry-run`, but it appears that the answer is no - or at least, it would be hard to take advantage of anything like that.

So, given the above requirement, we felt that the best approach is to make it less likely that this issue will occur, and to improve the documentation when it does occur.

There are three classes of untracked files that I can think of:

1. Files created in-source by various build processes

   For these, we should use `.gitignore` and `svn:ignore` to the fullest extent.

2. case directories that users often create in `cime/scripts`

   We should be able to have a `.gitignore` file in cime that ignores everything under scripts except for the things that are truly under version control there. This will carry some maintenance cost, but @mvertens and I felt this was a manageable cost.

3. Everything else - e.g., temporary files created by users

   We'll deal with this by improving the documentation, as @mvertens noted above.


Longer term (post-v1), it's possible that we could avoid some of these situations by saying: if (for example) we're trying to update cime, but mosart has untracked files, that shouldn't block the update of cime. i.e., untracked files should only block an update if you're actually trying to update the external containing the untracked files. However, this is likely more trouble than it's worth, since it would only get around the problem some of the time.

**Comment by billsacks** [comment]
Side-note: In doing some experimentation, I noticed a scenario where you can lose data with git, which gets potentially worse if you have a catch-all `.gitignore` file like what we're proposing for `cime/scripts`: If you have a file or directory in your working directory that is git-ignored, and then you checkout a branch where there is a file or directory with the same name that has been committed to the repository, then your ignored file will be overwritten in a non-recoverable manner. I haven't thought of a scenario where this is likely to be a problem in practice, but given enough time this may bite someone, sometime.

I still think the set of solutions in the above comment is the right way to go, but I wanted to make others aware of this caveat.

**Comment by bandre-ucar** [comment]
Given the potential for data loss, I think using git ignore should be rejected, and for cime we should just stop encouraging people to create tests and cases in cime/scripts.  Maybe even make the default to the source root.

**Comment by billsacks** [comment]
> Given the potential for data loss, I think using git ignore should be rejected, and for cime we should just stop encouraging people to create tests and cases in cime/scripts. Maybe even make the default to the source root.

Let's move this conversation to https://github.com/ESMCI/cime/issues/2182 since it's really a cime issue and that way we can get a wider range of viewpoints.

**Comment by gold2718** [comment]
I still think this is an issue that should be dealt with here.
For instance, we can add code that checks to see if there are untracked files which would be overwritten. A list of untracked files could be from:
```
git clean -X -n | cut -f3 -d' '
```
For the files in this list, the script could see if that file exists in the branch about to be checked out. A bash version would be something like:
```
for file in `git clean -X -n | cut -f3 -d' '`; do
    git cat-file -e <branch>:$file 2> /dev/null
    res=$?
    if [ $res -eq 0 ]; then
        echo "ERROR: File $file would be overwritten"
        exit(abort-self-destruct code)
    fi
done
```
where `<branch>` is the branch (or tag) to be checked out. This can take a little bit of time for something like CIME which creates quite a few `.pyc` and 'buildnmlc` files but I think it is worth it to make this tool more robust.

**Comment by billsacks** [comment]
@gold2718 -

**Note on the below comment:** I wrote this comment thinking that you were trying to find files that are untracked but _not_ in your `.gitignore`, which would therefore prevent git from doing a checkout. As I was about to post it, I realized that maybe you're referring to the problem I raised in https://github.com/NCAR/manage_externals/issues/51#issuecomment-355184643 . I'll keep this comment as-is, and will write a follow-up comment. So take this comment to be addressing the question: **Should we apply something like Steve's solution to do an up-front check of untracked files that are not in your `.gitignore`?**

@mvertens @bandre-ucar and I had a long discussion yesterday, which I'll summarize in a following comment. The short answer is that we decided to deal with this issue by allowing `checkout_externals` to run in the presence of untracked files; @mvertens and I felt that the downside risk of doing so was fairly minimal.

The idea of checking whether files would be overwritten sounds good in principle, though personally I feel that the amount we gain from this isn't worth the cost.

What do we gain? Another way to look at this is: What's the risk/cost of _not_ doing this check? As I'll elaborate on in a following comment, the risk/cost of not doing this check is that some externals would be in an updated state and some would not (with an error message issued informing the user of this and letting them know what needs to be done). To me, this isn't ideal, but isn't a major problem: it's not _that_ much worse than a state where _none_ of the externals have been updated (which is what you'd get if you do this check up-front).

What's the cost?

1. It will increase the runtime of `checkout_externals`, possibly substantially. On cheyenne (with its relatively slow file system), with a pretty dirty working copy of cime, I ran similar commands to the ones you gave, but used `git clean -n -d` (to find untracked files _not_ in `.gitignore`, and including directories; I just realized that maybe this is addressing a different problem from the one you meant to address). I found that these commands took roughly 30 sec to run (adding the `-d` seemed to be a large culprit here; without that, it was maybe 5 sec or so).

2. Extra complexity in the code

I don't feel that the gains justify the costs here, but I'm open to be overruled on this point, particularly if this check is only done when a given external is out of sync with `CESM.cfg` (e.g., you wouldn't do this check on cime if you were only going to update the cam external).

**Comment by billsacks** [comment]
@gold2718 Okay, now I'll address what I think you really meant (sorry about that!) - i.e., preventing the problem I pointed out in https://github.com/NCAR/manage_externals/issues/51#issuecomment-355184643

That problem does worry me. However, it's hard to argue that it's a problem caused by `manage_externals`, and so I'm also not sure that it's worth trying to solve that problem with `manage_externals`, given that it will add extra complexity and cost (i.e., time to run). In general, I don't know that we want to get in the business of having `manage_externals` contain workarounds for what I'd view as bugs in git.

**Comment by ekluzek** [comment]
@bandre-ucar @billsacks let me throw out another idea, based on what subversion does with externals updates. There are some times when subversion will copy the previous external and add a ".old" to the end of it, rather than overwriting that version. It also has some numbering scheme beyond the single ".old", but I'd have to check what it is.

We could adopt this same strategy, so your ".old" version has the untracked changes. I think since we essentially are trying to replicate the operation of subversion externals it does make sense that we should adopt the same protocol that it uses.

**Comment by billsacks** [comment]
@ekluzek At least with recent versions of subversion it appears that doing an `svn switch` will not overwrite the old version, but instead will put your directory into a state of conflict which you can then recover from. To me, that's preferable (easier to recover from) over creating a separate directory with `.old`. But do you see situations where you'd be better off having this `.old` directory than you would be if you just did an `svn switch`?

**Comment by billsacks** [comment]
@mvertens @bandre-ucar and I had a long meeting yesterday to discuss this issue. Here is a summary of this discussion for those interested (and so we can remember later why we chose the path we chose).

tl;dr: We decided to have `checkout_externals` proceed in the presence of untracked files, issuing an informative error message in the (somewhat rare) case when a given external cannot be updated because an untracked file in the sandbox conflicts with an incoming tracked file. (This is solution (3) in the below list, and is the solution proposed in my original comment.) @mvertens and I felt that this was better than the status quo; @bandre-ucar cast a dissenting vote, preferring to stick with the current behavior where `checkout_externals` does not proceed in the presence of untracked files - but he was willing to go along with the change.

Details:

The basic problem is: What should happen when you try to rerun `checkout_externals` when you have untracked files in your sandbox?

We first noted that this is only an issue when someone has already checked out the model, then wants to update it to new versions of the externals. We can see two use cases where this will often occur:

- Check out a CESM version, then bring in different versions of some components

- When doing development on a standalone checkout of (say) CTSM: after merging the latest version of master into your branch, rerunning `checkout_externals` to get any updates that came along with latest master.

We came up with six possible solutions, though some of them are combinations of others:

0. (Not really a solution, just included for completeness) Say that updating an existing sandbox is not a supported use of `checkout_externals`: it is only designed to give you an initial checkout, at which point managing your sandbox is up to you.

1. Status quo: `checkout_externals` checks for untracked files (along with its other checks); if it finds any, then it aborts without updating anything.

2. (1) combined with a much more aggressive use of `.gitignore` files, including ignoring cases in cime/scripts. However, this was largely rejected in conversation in https://github.com/ESMCI/cime/issues/2182 and elsewhere. (But note that we do still want to add obvious things to `.gitignore` files, such as `.pyc` files created by the build.)

3. Pay no attention to untracked files. If the presence of untracked files ends up causing a problem in a git or svn command, then `checkout_externals` will report an error to the user, letting them know which external(s) need to be fixed. Note that this could result in some externals being updated and others not.

4. Do a `dry-run` of the checkout using any supported flags in git / svn; if the dry-run indicates there would be problems, then `checkout_externals` aborts without updating anything. This may be possible with svn, but seems to be very messy for git (because `--dry-run` isn't supported for all necessary sub-commands).

5. Variation on (1): Only allow untracked files to stop `checkout_externals` if those untracked files are in externals that need to be updated. e.g., if you have only changed the mosart tag and only have untracked files in cime, then those untracked files won't stop the mosart update. (The feeling was that _modified_ tracked files should _probably_ still stop the update even if they're in components that aren't being updated.)

6. Either (1) or (5) with an optional flag (e.g., `--allow-untracked`) allowing users to get the behavior of (3). 

Of these, we gave most consideration to (1), (3) and (6) (though figuring that, if we went with (1), we'd consider (5)).

**(1) Status quo: `checkout_externals` checks for untracked files (along with its other checks); if it finds any, then it aborts without updating anything.**

Where will it fail? If there are any untracked files in any externals' directories in your sandbox.

How often will it fail? Even after addressing low-hanging fruit like `.pyc` files, we still expect it to fail somewhat often. Examples of untracked files are:

- cases in cime/scripts - hard to deal with

- misc. files (notes, etc.) - very hard to deal with up-front, though possible for the user to move these files out of the way before running `checkout_externals`

- stdout from jobs (e.g., from scripts_regression_tests)

What's needed to recover? One of the following options:

- Move / remove offending files and rerun `checkout_externals` (painful for cases)

- Manually run the necessary git/svn commands to update each external

- Option (6)

- Create a new sandbox rather than trying to update your existing sandbox

**(3) Pay no attention to untracked files. If the presence of untracked files ends up causing a problem in a git or svn command, then `checkout_externals` will report an error to the user, letting them know which external(s) need to be fixed. Note that this could result in some externals being updated and others not.**

Where will it fail? If there are untracked files in your sandbox, but the branch/tag you're updating to has that file checked in in the repository. This can lead to two types of failures:

- git/svn says "I can't do this checkout operation". It seems that git takes this approach

- git/svn tries to do the checkout operation but fails. It seems that svn takes this approach, leaving you in a "conflict" state.

In either case, you'll be left in a state where some externals have been checked out to the new branch/tag that you requested, but at least one is left either in the old state (git) or in a conflict state (svn), which needs to be resolved.

How often will this occur? @bandre-ucar felt it could occur somewhat frequently in the midst of major refactorings - e.g., where you have renamed a file, but then accidentally save the file with the old name from your editor, then switch to a different branch where that file was present... he experiences issues like that a few times per week when in the midst of major refactoring. I felt that the problem would generally occur (much) less often: maybe closer to once a month for developers and less often for casual users. (But @bandre-ucar pointed out that this could still occur for casual users, e.g., if they copy a new source file from someone else's sandbox to their own, then later switch to point to the branch from which that file originated.)

What's needed to recover? First, we need to convey the message CLEARLY to the user about what happened and how to recover. Recovery involves:

- For git: Remove offending untracked files then either rerun `checkout_externals` or do a manual `git checkout`

- For svn: Resolve the svn conflict - exact steps TBD

Risks if you don't recover properly:

- For git: At least one external doesn't match its entry in `CESM.cfg`

- For svn: An external is in a conflicted state

Importantly (to me anyway): We could not envision a situation where this would result in data loss or leaving you in an un-recoverable state.

**We also discussed a possible modification - solution (3a): Rather than having `checkout_externals` bail as soon as one externals update raises an error, instead have it move along to update all other externals. So the only ones that are in a non-updated state are the ones that actually raised an error in doing the update.** My sense was that we liked this idea, as long as the cost of implementing it isn't too great.

**(6) Either (1) or (5) with an optional flag (e.g., `--allow-untracked`) allowing users to get the behavior of (3).**

This has the advantage that different people can make their own choices about which they prefer. However, we rejected this because (i) it would create additional complexity, and (ii) in some senses, it comes with the downsides of both (1) and (3): Users will still experience the frustrations of (1) until they learn to add this flag, and users who do add this flag will still suffer the risks of (3).

**Our decision**

We felt that (3) leads to a somewhat riskier situation when it occurs, since if the user doesn't notice the error message, they'll have a sandbox in an inconsistent state, with possibilities for frustration or creating a case differently from how you intended. However, it should occur much less frequently than (1). @mvertens and I felt that (1) would cause substantial frustration for users, many of whom are accustomed to putting miscellaneous files in their sandbox. In addition, (1) also creates the possibility for creating a case differently from how you intended (if you don't pay attention to the error message), since your entire sandbox is left in the previous state. So @mvertens and I felt that (3) was preferable overall. @bandre-ucar still preferred (1) but was willing to go with (3).
